--- a/drivers/net/ethernet/intel/e1000e/e1000.h	2017-06-29 22:55:48.000000000 +0800
+++ b/drivers/net/ethernet/intel/e1000e/e1000.h	2018-08-29 13:20:07.333073376 +0800
@@ -133,7 +133,8 @@
 	board_pchlan,
 	board_pch2lan,
 	board_pch_lpt,
-	board_pch_spt
+	board_pch_spt,
+	board_pch_cnp
 };
 
 struct e1000_ps_page {
@@ -505,6 +506,7 @@
 extern const struct e1000_info e1000_pch2_info;
 extern const struct e1000_info e1000_pch_lpt_info;
 extern const struct e1000_info e1000_pch_spt_info;
+extern const struct e1000_info e1000_pch_cnp_info;
 extern const struct e1000_info e1000_es2_info;
 
 void e1000e_ptp_init(struct e1000_adapter *adapter);
